### DISCUSSION TITLE
Installing new LSPs?

### DISCUSSION THREAD

**Discussion by damccull** [nix-community/kickstart-nix.nvim#31] [Q&A] [answered]
Hello, do you have an example of how to install and configure new LSPs? I want to get rust and rust-analyzer working, and I'm having a hard time understanding how to do that. In kickstart.nvim, I would use lazy with lsp-config and mason to install the LSPs I want and to load them lazily when a specific file type is opened. How do I do that, or as much of that as possible, with this version?

**Comment by mrcjkb** [comment]
Hey again :wave: 

With Nix, there's different ways to add LSP servers.

- One way is by adding them to [`extraPackages`](https://github.com/nix-community/kickstart-nix.nvim/blob/52b723c5534bf8227d7123adf6f055ee211c7338/nix/neovim-overlay.nix#L86)
  This will add them as part of your Neovim package. I personally prefer not to do that for most servers.
- Another option is to add them to Nix `devShells`. I prefer to do this, as it ensures I have the correct server versions for the project I'm working on.

**Reply by damccull** [reply]
Would you mind giving an example of the devshell version?

**Reply by mrcjkb** [reply, marked as answer]
Sure. this repo's flake [has one](https://github.com/nix-community/kickstart-nix.nvim/blob/37bc6855bde7ee36bf4c08dac0c24009f30d47c2/flake.nix#L46).

You can enter it with `nix develop`.

**Comment by damccull** [comment]
Oh. Neat. I did that actually. So this shell is what loads the actual lsp itself, and I still need to ensure my plugins match up so that I can use it all, right? 

And that's what the derivations are for, if I'm understanding correctly. I load a dev shell then launch the correct derivation to match it, with this flakes current architecture?

**Reply by mrcjkb** [reply]
> and I still need to ensure my plugins match up so that I can use it all, right?

Yes, although I use Neovim's built-in LSP client API directly, instead of lspconfig.
For language servers that have off-spec features (like rust-analyzer), I use a dedicated plugin. 

> And that's what the derivations are for, if I'm understanding correctly. I load a dev shell then launch the correct derivation to match it, with this flakes current architecture?

That's entirely up to you 😄 
It's one way you can do it. If you have plugins that have a heavy impact, but you don't need them all the time, it could be useful to create a derivation without them.

If you look at the `ftplugin` directory (`:h ftplugin`), that's where I set LSP clients up. As opposed to lspconfig, this has zero impact on startup time.
I just check if the language server executable exists and abort if it doesn't. That way I don't need to worry about which files to exclude.

Personally, I just have two derivations: One with all my plugins in the nix store and one with the plugins I maintain checked out locally.

**Comment by damccull** [comment]
Oh, rust-analyzer needs a separate plug-in to work at all? Or just to use some extras it may have? What plugin do you recommend for it?

**Reply by mrcjkb** [reply]
Just for additional features that aren't part of the official LSP spec. For everything else it works fine with Neovim's built-in API or lspconfig.

I maintain [a plugin](https://github.com/mrcjkb/rustaceanvim) for it.

**Comment by damccull** [comment]
Hey that's cool. I've actually seen that one before and tried, unsuccessfully, to figure it out. I'll have to give it a try with this setup now as I won't need to figure out how to integrate it with lspconfig, I think.
